src/Common/src/TypeSystem/Interop/InteropStateManager.cs

```diff
@@ -0,0 +1,205 @@
+            return _structMarshallingThunkHashtable.GetOrCreateValue(methodKey);
+        }
+
+
+        private struct NativeStructTypeHashtableKey
+        {
+            public readonly MetadataType ManagedStructType;
+            public readonly ModuleDesc OwningModule;
+
+            public NativeStructTypeHashtableKey(MetadataType type, ModuleDesc owningModule)
+            {
+                ManagedStructType = type;
+                OwningModule = owningModule;
+            }
+        }
+
+        private class NativeStructTypeHashtable : LockFreeReaderHashtable<NativeStructTypeHashtableKey, NativeStructType>
+        {
+            protected override int GetKeyHashCode(NativeStructTypeHashtableKey key)
+            {
+                return key.ManagedStructType.GetHashCode();
+            }
+
+            protected override int GetValueHashCode(NativeStructType value)
+            {
+                return value.ManagedStructType.GetHashCode();
+            }
+
+            protected override bool CompareKeyToValue(NativeStructTypeHashtableKey key, NativeStructType value)
+            {
+                return Object.ReferenceEquals(key.ManagedStructType, value.ManagedStructType) &&
+                    Object.ReferenceEquals(key.OwningModule, value.Module);
+            }
+
+            protected override bool CompareValueToValue(NativeStructType value1, NativeStructType value2)
+            {
+                return Object.ReferenceEquals(value1.ManagedStructType, value2.ManagedStructType) &&
+                    Object.ReferenceEquals(value1.Module, value2.Module);
+            }
+
+            protected override NativeStructType CreateValueFromKey(NativeStructTypeHashtableKey key)
+            {
+                return new NativeStructType(key.OwningModule, key.ManagedStructType, _interopStateManager);
+            }
+
+            private readonly InteropStateManager _interopStateManager;
+
+            public NativeStructTypeHashtable(InteropStateManager interopStateManager)
+            {
+                _interopStateManager = interopStateManager;
+            }
+        }
+
+        private struct StructMarshallingThunkKey
+        {
+            public readonly MetadataType ManagedType;
+            public readonly TypeDesc OwningType;
+            public readonly StructMarshallingThunkType ThunkType;
+
+            public StructMarshallingThunkKey(MetadataType type, TypeDesc owningType, StructMarshallingThunkType thunkType)
+            {
+                ManagedType = type;
+                OwningType = owningType;
+                ThunkType = thunkType;
+            }
+        }
+
+        private class StructMarshallingThunkHashTable : LockFreeReaderHashtable<StructMarshallingThunkKey, StructMarshallingThunk>
+        {
+            protected override int GetKeyHashCode(StructMarshallingThunkKey key)
+            {
+                return key.ManagedType.GetHashCode() | (int)key.ThunkType;
```

Comment:
Bitwise or is a terrible hash combining function because it tends to result in a lot of 1 bits...